### PR TITLE
Fixed #31649 -- Added support for covering exclusion constraints on PostgreSQL 12+

### PR DESCRIPTION
**Pull request by hannseman**
Ticket: https://code.djangoproject.com/ticket/31649

### CODE REVIEW

File: tests/postgres_tests/test_constraints.py [hunk 1]
```diff
@@ -417,3 +462,45 @@ def test_range_adjacent_initially_deferred(self):
...
+
+    @skipUnlessDBFeature('supports_covering_gist_indexes')
+    def test_range_adjacent_include(self):
+        constraint_name = 'ints_adjacent_include'
+        self.assertNotIn(constraint_name, self.get_constraints(RangesModel._meta.db_table))
+        constraint = ExclusionConstraint(
+            name=constraint_name,
+            expressions=[('ints', RangeOperators.ADJACENT_TO)],
+            include=['decimals', 'ints'],
+            index_type='gist'
```
Comment (felixxm) [inline tests/postgres_tests/test_constraints.py:474]: Add trailing comma.

File: django/contrib/postgres/constraints.py [hunk 2]
```diff
@@ -107,15 +115,17 @@ def __eq__(self, other):
...
 
     def __repr__(self):
-        return '<%s: index_type=%s, expressions=%s%s%s>' % (
+        return '<%s: index_type=%s, expressions=%s%s%s%s>' % (
             self.__class__.__qualname__,
             self.index_type,
             self.expressions,
             '' if self.condition is None else ', condition=%s' % self.condition,
             '' if self.deferrable is None else ', deferrable=%s' % self.deferrable,
+            '' if not self.include else ', include=%s' % str(self.include),
```
[the remark this answers is not in the document]
Comment (pope1ni) [reply django/contrib/postgres/constraints.py:130]: Also use `%r` and not `repr()`.

[the remark this answers is not in the document]
Comment (felixxm) [reply django/contrib/postgres/constraints.py:130]: `include` is a tuple, that's why we need `repr()`.

File: django/contrib/postgres/constraints.py [hunk 3]
```diff
@@ -39,10 +39,14 @@ def __init__(
             raise ValueError(
                 'ExclusionConstraint.deferrable must be a Deferrable instance.'
             )
+        if not isinstance(include, (type(None), list, tuple)):
+            raise ValueError('ExclusionConstraint.include must be a list or tuple.')
+
```
[the remark this answers is not in the document]
Comment (hannseman) [reply django/contrib/postgres/constraints.py:46]: Good idea, I added a `check_supported`-method to do the latter check, much like `PostgresIndex. check_supported`.

File: django/contrib/postgres/constraints.py [hunk 4]
```diff
@@ -67,10 +71,12 @@ def constraint_sql(self, model, schema_editor):
         compiler = query.get_compiler(connection=schema_editor.connection)
         expressions = self._get_expression_sql(compiler, schema_editor.connection, query)
         condition = self._get_condition_sql(compiler, schema_editor, query)
+        include = [model._meta.get_field(field_name).column for field_name in self.include]
         return self.template % {
             'name': schema_editor.quote_name(self.name),
             'index_type': self.index_type,
             'expressions': ', '.join(expressions),
+            'include': ' INCLUDE (%s)' % ', '.join(include) if include else '',
```
[the remark this answers is not in the document]
Comment (hannseman) [reply django/contrib/postgres/constraints.py]: Haha wow I don't know how I forgot that method. Thanks 👍

**Comment by felixxm** [comment]
@hannseman Thanks :+1: I pushed minor edits.

**Comment by hannseman** [comment]
@felixxm Thanks for the review and fixups!

**Pull request merged**